File: lclang/src/main/java/lclang/statements/Statement.java
package lclang.statements;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.LCRootExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;

public abstract class Statement {
    public final int line;

    public Statement(int line) {
        this.line = line;
    }

    public abstract Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException;

    public Caller getCaller(Caller caller) {
        return new Caller(caller.root, line, caller);
    }
    public Caller getCaller(LCRootExecutor root) {
        return new Caller(root, line, null);
    }
}

File: lclang/src/main/kotlin/lclang/statements/Statement.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.Value

abstract class Statement(val line: Int, val column: Int) {
    abstract fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value

    fun getCaller(caller: Caller) = Caller(caller.root, line, column, caller)
}